### PR TITLE
Refactor prometheus into StatsCollector interface

### PR DESCRIPTION
**Pull request by comstud**
This adds a StatsCollector interface so that stats implementations can be pluggable. This has the following benefits beyond it being possible for someone to implement something other than prometheus:

* We can now create and default a 'no-op' collector. As things are today, if prometheus is disabled, all of the calls are still doing some amount of work. This PR makes stats collection calls truly no-ops.

* The potential for bugs is reduced. Right now it is very easy to forget a label value, because WithLabelValues() takes a variable number of values. If the number of values does not match the number that were registered, a panic will occur. If we have method signatures with the correct arguments, bugs are limited to the stats collection code, itself.

* Since the stats collector is now pluggable, we can start to test that stats calls are being made when we expect them to be made by creating mock stats collectors in tests and injecting them. This PR does not attempt to add any such tests, yet, however.

### CODE REVIEW

**Comment by comstud** [comment]
I am running this with prometheus enabled and no problems. No changes to graphs.

**Comment by Fabio1988** [comment]
Do you still the same import formatting tool like IntelliJ GoLand? Seems either it was wrong before or after your changes :)

File: main.go [hunk 1]
```diff
@@ -236,7 +240,24 @@ func main() {
 	// Start the web server.
 	gin.SetMode(gin.ReleaseMode)
 	r := gin.New()
-	external.InitPrometheus(r) // init prometheus if enabled
+
+	// choose the statsCollector we will use.
+	if cfg.Prometheus.Enabled {
```
Comment (Fabio1988) [inline main.go:245]: Could we move that into a private helper function? To make main flow more clean? Maybe only personal preference :sweat_smile:

Comment (comstud) [reply main.go]: 👍 I moved it into stats_collector

**Comment by comstud** [comment]
> Do you still the same import formatting tool like IntelliJ GoLand? Seems either it was wrong before or after your changes :)

We should figure out what we want to do. :) Jabes said he did like separating the 3rd party dependencies.

A common organization technique is this:

```
<stdlib>

<3rd-party>

<local-to-your-project-packages>
```

**Comment by Fabio1988** [comment]
> > Do you still the same import formatting tool like IntelliJ GoLand? Seems either it was wrong before or after your changes :)
> 
> We should figure out what we want to do. :) Jabes said he did like separating the 3rd party dependencies.
> 
> A common organization technique is this:
> 
> ```
> <stdlib>
> 
> <3rd-party>
> 
> <local-to-your-project-packages>
> ```

Yeah I like it as well, you use `go fmt`?

**Comment by comstud** [comment]
> > > Do you still the same import formatting tool like IntelliJ GoLand? Seems either it was wrong before or after your changes :)
> > 
> > 
> > We should figure out what we want to do. :) Jabes said he did like separating the 3rd party dependencies.
> > A common organization technique is this:
> > ```
> > <stdlib>
> > 
> > <3rd-party>
> > 
> > <local-to-your-project-packages>
> > ```
> 
> Yeah I like it as well, you use `go fmt`?

vim-go is actually using 'goimports' to find and add imports for me automatically. And then I believe goimports is calling 'go-fmt' afterwards.

go-fmt doesn't seem to care too much about how you organize the imports into different sections/blocks. If you separate things into blocks, it seems to keep them that way, but sorts each block.

According to goimports docs, '-l golbat/' *should* make it move those imports to be after the 3rd party imports, but I can't get that to work automatically. I have to manually move 'golbat/' imports to be after 3rd party in its own block. But once I do that one time, everything is sorted and maintained.

I suspect if someone did a PR to re-organize them all, vim would stop moving them around. And I suspect you can figure out how to get Jetbrains to work, also.

I don't mind doing the work and submitting the PR to reorganize them all, if you want. It looks like I goofed up in at least one of my PRs and put 'golbat' between 'stdlib' and '3rd party'. 'golbat' should be *after* 3rd party. Oops :)

**Comment by jfberry** [comment]
> I don't mind doing the work and submitting the PR to reorganize them all, if you want. It looks like I goofed up in at least one of my PRs and put 'golbat' between 'stdlib' and '3rd party'. 'golbat' should be _after_ 3rd party. Oops :)

I think these are the Goland settings which reflect what goimports would do? (we should probably not be too distracted by this in this PR!)

image

**Review by Fabio1988 — APPROVED** [review]

File: stats_collector/prometheus.go [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+	}
+
+	verifiedPokemonTTL.WithLabelValues(area.String(), seenTypeStr, above30).Add(float64(remainingTtlMin))
+	verifiedPokemonTTLCounter.WithLabelValues(area.String(), seenTypeStr, above30).Inc()
+}
+
+func (col *promCollector) UpdateRaidCount(areas []geo.AreaName, raidLevel int64) {
+	processed := make(map[string]bool)
+	for _, area := range areas {
+		if !processed[area.String()] {
```
Comment (jfberry) [inline stats_collector/prometheus.go:342]: I think this is cut & paste code, but the duplicate call to .String() here may not be able to be optimised by the compiler (but we know it is the same value)

Comment (comstud) [reply stats_collector/prometheus.go]: Yeah, those functions were copied as-is and I didn't pay much attention to them. I agree with only calling String() once. This has been updated now in all of the places...

Comment (comstud) [reply stats_collector/prometheus.go]: https://github.com/UnownHash/Golbat/compare/be15459f094400fe5ea9bdc5b1ef238238a71542..a23d1a5b9b56e7caada81816748d55e8f5da1d2b

**Review by jfberry — APPROVED** [review]
Just for @lenisko to approve then can be merged

**Review by lenisko — APPROVED** [review]
Sorry it took so long. I don't see any problems, green light from me. Good work!

**Comment by lenisko** [comment]
@comstud Actually... when I run it I got this little fella
```
Oct 26 13:38:52 golbat[2523985]: 2023/10/26 13:38:52 [Recovery] 2023/10/26 - 13:38:52 panic recovered:
Oct 26 13:38:52 golbat[2523985]: runtime error: invalid memory address or nil pointer dereference
Oct 26 13:38:52 golbat[2523985]: /usr/local/go/src/runtime/panic.go:261 (0x4541f7)
Oct 26 13:38:52 golbat[2523985]:         panicmem: panic(memoryError)
Oct 26 13:38:52 golbat[2523985]: /usr/local/go/src/runtime/signal_unix.go:861 (0x4541c5)
Oct 26 13:38:52 golbat[2523985]:         sigpanic: panicmem()
Oct 26 13:38:52 golbat[2523985]: /home/user/library/golbat/routes.go:297 (0xf87033)
Oct 26 13:38:52 golbat[2523985]:         Raw: statsCollector.IncRawRequests("ok", "")
Oct 26 13:38:52 golbat[2523985]: /home/user/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xe4b579)
Oct 26 13:38:52 golbat[2523985]:         (*Context).Next: c.handlers[c.index](c)
Oct 26 13:38:52 golbat[2523985]: /home/user/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/recovery.go:102 (0xe4b567)
Oct 26 13:38:52 golbat[2523985]:         CustomRecoveryWithWriter.func1: c.Next()
Oct 26 13:38:52 golbat[2523985]: /home/user/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xe3f54a)
Oct 26 13:38:52 golbat[2523985]:         (*Context).Next: c.handlers[c.index](c)
Oct 26 13:38:52 golbat[2523985]: /home/user/go/pkg/mod/github.com/!depado/ginprom@v1.7.11/prom.go:309 (0xe6e504)
Oct 26 13:38:52 golbat[2523985]:         GetStatsCollector.(*Prometheus).Instrument.func6: c.Next()
Oct 26 13:38:52 golbat[2523985]: /home/user/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xe4a45a)
Oct 26 13:38:52 golbat[2523985]:         (*Context).Next: c.handlers[c.index](c)
Oct 26 13:38:52 golbat[2523985]: /home/user/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:620 (0xe4a0ed)
Oct 26 13:38:52 golbat[2523985]:         (*Engine).handleHTTPRequest: c.Next()
Oct 26 13:38:52 golbat[2523985]: /home/user/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:576 (0xe49c1c)
Oct 26 13:38:52 golbat[2523985]:         (*Engine).ServeHTTP: engine.handleHTTPRequest(c)
Oct 26 13:38:52 golbat[2523985]: /usr/local/go/src/net/http/server.go:2938 (0xb5df4d)
Oct 26 13:38:52 golbat[2523985]:         serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
Oct 26 13:38:52 golbat[2523985]: /usr/local/go/src/net/http/server.go:2009 (0xb59e33)
Oct 26 13:38:52 golbat[2523985]:         (*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
Oct 26 13:38:52 golbat[2523985]: /usr/local/go/src/runtime/asm_amd64.s:1650 (0x470fa0)
Oct 26 13:38:52 golbat[2523985]:         goexit: BYTE        $0x90        // NOP
Oct 26 13:38:52 golbat[2523985]:
Oct 26 13:38:52 golbat[2523985]: panic: runtime error: invalid memory address or nil pointer dereference
Oct 26 13:38:52 golbat[2523985]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0xf81c87]
Oct 26 13:38:52 golbat[2523985]: goroutine 30 [running]:
Oct 26 13:38:52 golbat[2523985]: main.decode({0x17c9408, 0xc00087bc00}, 0xb, 0xc01f267ec0)
Oct 26 13:38:52 golbat[2523985]:         /home/user/library/golbat/main.go:417 +0xc47
Oct 26 13:38:52 golbat[2523985]: main.Raw.func7()
Oct 26 13:38:52 golbat[2523985]:         /home/user/library/golbat/routes.go:288 +0xf7
Oct 26 13:38:52 golbat[2523985]: created by main.Raw in goroutine 28
Oct 26 13:38:52 golbat[2523985]:         /home/user/library/golbat/routes.go:256 +0xbac
Oct 26 13:38:53 systemd[2373]: golbat.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct 26 13:38:53 systemd[2373]: golbat.service: Failed with result 'exit-code'.
Oct 26 13:38:53 systemd[2373]: golbat.service: Consumed 9.339s CPU time.
```

**Comment by Fabio1988** [comment]
Might be related to last commit :/

**Comment by lenisko** [comment]
Reverted it on `main`. Lets wait for @comstud

**Pull request merged**